biggraphite/drivers/elasticsearch.py

```diff
@@ -399,21 +472,78 @@ def delete_metric(self, name):
     def delete_directory(self, name):
         self._directory_names.remove(name)
 
-    @staticmethod
-    def __glob_names(names, glob):
-        results = bg_glob.glob(names, glob)
-        results.sort()
-        return results
+    def _search_metrics_from_glob(self, glob):
+        search = elasticsearch_dsl.Search()
+        search = search.using(self.client).index("%s*" % self._index_prefix).source('name')
+
+        components = self.__glob_parser.parse(glob)
+
+        # Handle glob with globstar(s).
+        globstars = components.count(bg_glob.Globstar())
+        if globstars:
+            name_regexp = "\\.".join([_parse_regexp_component(c) for c in components])
+            return True, search.filter('regexp', **{"name": name_regexp})
+
+        # Handle fully defined glob.
+        if self.__glob_parser.is_fully_defined(components):
+            return False, search.filter(
+                'term', **{"name": ".".join(self._components_from_name(glob))})
+
+        # Handle all other use cases.
+        for i, c in enumerate(components):
+            if len(c) == 1:
+                filter_type, value = parse_simple_component(c)
+            else:
+                filter_type, value = parse_complex_component(c)
+
+            if filter_type:
+                search = search.filter(filter_type, **{"p%d.keyword" % i: value})
+        return False, search
 
     def glob_metric_names(self, glob):
         """See the real Accessor for a description."""
         super(_ElasticSearchAccessor, self).glob_metric_names(glob)
-        return iter(self.__glob_names(self._metric_names, glob))
+        if glob == "":
+            return []
+
+        has_globstar, search = self._search_metrics_from_glob(glob)
+        if has_globstar:
+            search = search.filter('range', depth={'gte': glob.count(".")})
+        else:
+            search = search.filter('term', depth=glob.count("."))
+        search = search.extra(from_=0, size=MAX_QUERY_SIZE)
+
+        # TODO (t.chataigner) try to move the sort in the ES search.
+        log.debug(search.to_dict())
+        results = [h.name for h in search.execute()]
+        results.sort()
+        return iter(results)
 
     def glob_directory_names(self, glob):
         """See the real Accessor for a description."""
         super(_ElasticSearchAccessor, self).glob_directory_names(glob)
-        return iter(self.__glob_names(self._directory_names, glob))
+        if glob == "":
+            return []
+
+        has_globstar, search = self._search_metrics_from_glob(glob)
+        if has_globstar:
+            # FIXME (t.chataigner) Add a log or raise exception.
```

Comment:
change FIXME in TODO if you don't want to implement it now

Reply:
Done